### PR TITLE
listeners: add unified matcher for filter chains

### PR DESCRIPTION
**Pull request by kyessenov**
Commit Message: Add unified matcher for network streams, as a replacement for filter chain match. See previous discussion in https://github.com/envoyproxy/envoy/pull/18871
Risk Level: low (requires opt-in)
Testing: unit, integration
Docs Changes: yes
Release Notes: yes
Fixes: #3411 #18685

### CODE REVIEW

**Comment by kyessenov** [comment]
Working on additional test coverage but the implementation is ready.

**Comment by kyessenov** [comment]
cc @mattklein123

**Comment by kyessenov** [comment]
I verified most listener tests work with matchers. There are gaps in the matching framework that can be solved separately:
* we didn't implement `on_no_match` for `custom_matchers`;
* we need better ALPN list matching;
* we need to finish server name matching (https://github.com/cncf/xds/pull/24);
* it's overly verbose due to the use of `TypedExtensionsConfig`; little things like https://github.com/cncf/xds/pull/28 can go a long way.

File: api/envoy/config/listener/v3/listener.proto
Comment (mattklein123) [inline api/envoy/config/listener/v3/listener.proto:192]: A few doc notes:
1) We will need a release note.
2) I feel strongly that we have to have a) documentation for each supported matcher type and b) an auto generated list of all of the registered matcher types that will work.

I apologize in advance for putting this on, but we are going to need to sort out some of the documentation issues with putting protos in the xDS repo. We will need to make sure we are running protodoc on that repo, and also will need to figure out a way to wire up that repo to our extension category documentation tooling. If you need more pointers let me know but I might advise contacting @phlax on Slack and scheduling a short meeting with him to come up with a plan here. Again, I'm sorry to put this on you, but we need to figure this out because we will accrue significant documentation debt by not doing this, and the more protos that go into the xDS repo the worse this will get. Thank you in advance.

Separately, thoughts on marking this WiP for some period of time just so we can get some bake on this before marking it production ready and locked for API and security?

File: source/server/filter_chain_manager_impl.cc
Comment (mattklein123) [inline source/server/filter_chain_manager_impl.cc:259]: nit: create_address_vector

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Very nice, pretty impressive how simple this ends up being in the end once the other changes are done. I would definitely like to get @snowp opinion on this so will assign him also.

In the meantime my main concern is around documentation so can we work on sorting that out? Thank you!

/wait

File: api/envoy/config/listener/v3/listener.proto
Comment (kyessenov) [reply api/envoy/config/listener/v3/listener.proto:192]: Ack on making this WIP. I'll tag the protos as such and add a release note about experimental status for now. For documentation of cncf/xds, I'll ping @phlax and probably @markdroth. For matcher documentation, we should be able to figure out a table for inputs (TCP vs HTTP), and a list of all matchers using the existing extension categories.

File: source/server/filter_chain_manager_impl.cc
Comment (kyessenov) [reply source/server/filter_chain_manager_impl.cc:259]: This is actually a lambda (this code is moved as-is).

File: source/server/filter_chain_manager_impl.h
Comment (lambdai) [inline source/server/filter_chain_manager_impl.h:374]: Can you move this field close to `matcher_`?

Basically, one of below is activated, never both
1. destination_ports_map_, or
2. matcher_ + filter_chains_by_name_

File: source/server/filter_chain_manager_impl.cc
Comment (lambdai) [inline source/server/filter_chain_manager_impl.cc:574]: Since filter_chains_by_name_ is immutable after the listener build,
I am thinking about the result contains the reference to the filter chain, rather than the string name.

It might be an early optimization.

File: source/server/filter_chain_manager_impl.cc
Comment (lambdai) [reply source/server/filter_chain_manager_impl.cc:259]: I believe this green block is not modified other than the indent.
However, github UI fails to render the diff as we expected...

**Review by lambdai — COMMENTED** [review]
This is awesome! I have very a few early comment though

File: source/server/filter_chain_manager_impl.h
Comment (kyessenov) [reply source/server/filter_chain_manager_impl.h:400]: Thanks, this is a mistake, got confused by `OnMatch` shared ptr.

File: source/server/filter_chain_manager_impl.h
Comment (kyessenov) [reply source/server/filter_chain_manager_impl.h:374]: Done.

File: source/server/filter_chain_manager_impl.cc
Comment (kyessenov) [reply source/server/filter_chain_manager_impl.cc:568]: Done.

File: source/server/filter_chain_manager_impl.cc
Comment (kyessenov) [reply source/server/filter_chain_manager_impl.cc:574]: Good idea. Let me try add a context for the actions with the mapping.

**Comment by kyessenov** [comment]
@snowp Would you like to take a look at this?

File: api/envoy/config/listener/v3/listener.proto
Comment (zhxie) [inline api/envoy/config/listener/v3/listener.proto:139]: We still have an `on_no_match` field in the matcher, do you think it would be better to omit `default_filter_chain` together with `filter_chain_match`?

File: api/envoy/config/listener/v3/listener.proto
Comment (phlax) [inline api/envoy/config/listener/v3/listener.proto:141]: preferably use a `literalinclude` here and separate this to a yaml file

that way it will be checked both for validity and that a server can be started with the config - and will also give end users a working config to start from

File: api/envoy/config/listener/v3/listener.proto
Comment (kyessenov) [reply api/envoy/config/listener/v3/listener.proto:139]: Yeah, the default filter chain is unnecessary with the unified matcher. We can outright reject that has both set. Would that be cleaner?

Comment (zhxie) [reply api/envoy/config/listener/v3/listener.proto:139]: Sure. I am just a little bit confused by `the default filter chain is selected`.

Comment (kyessenov) [reply api/envoy/config/listener/v3/listener.proto]: Updated and simplified.

File: api/envoy/config/listener/v3/listener.proto
Comment (kyessenov) [reply api/envoy/config/listener/v3/listener.proto]: Done..

**Comment by kyessenov** [comment]
Added docs and a few examples. Tests are run with both legacy and new matchers except for a couple that need wildcard server name matching.

https://storage.googleapis.com/envoy-pr/90925f2/docs/intro/arch_overview/advanced/matching/matching_listener.html

File: docs/root/intro/arch_overview/advanced/matching/matching_listener.rst
Comment (phlax) [inline docs/root/intro/arch_overview/advanced/matching/matching_listener.rst:33]: if we are going to include `linenos` and the sample doesnt start from the first line,  then i think we also need to set `lineno-start` to the first line

it would also be good to include a caption so users can get the entire file:

```
:caption: :download:`listener_tls.yaml <_include/listener_tls.yaml>`
```

Comment (phlax) [reply docs/root/intro/arch_overview/advanced/matching/matching_listener.rst:33]: same below

**Review by mattklein123 — DISMISSED** [review]
Thanks, amazing work. I'm good with the code changes. I'm going to defer to @phlax for final merge once he is satisfied with the docs. It's super important that we do a good job documenting this and I really appreciate all the work you have put in here, it is really valuable for the community. Thank you!!!

**Comment by kyessenov** [comment]
Thank you all for the review!

File: docs/root/intro/arch_overview/advanced/matching/matching_listener.rst
Comment (kyessenov) [reply docs/root/intro/arch_overview/advanced/matching/matching_listener.rst:32]: Done. I removed the line numbers, don't seem essential.

File: docs/root/intro/arch_overview/advanced/matching/_include/listener_tls.yaml
Comment (howardjohn) [inline docs/root/intro/arch_overview/advanced/matching/_include/listener_tls.yaml:50]: Maybe I am missing some context, but could this be simplified to just have the `name`? it seems redundant to have the same info in two places and I worry if this may impact XDS sizes

File: source/server/filter_chain_manager_impl.cc
Comment (howardjohn) [inline source/server/filter_chain_manager_impl.cc:575]: nit: Currently a failed match is indistinguishable from the action FilterChain not being found (with default_filter_chain unset), even at debug log level. It may be helpful to have some sort of logging to help distinguish these, which is often user request error (sending wrong request and failing to match) vs config/control plane error (failed to send correct Filter Chain to go with the match)

Comment (howardjohn) [reply source/server/filter_chain_manager_impl.cc:575]: Scope creeping this a ton, for development purposes it could be nice to have some sort of debugging info on where in the tree we got. For some especially deep matching trees getting a binary "no match" can be pretty hard to decipher. It would be really cool if there was some sort of log like "match port,protocol,ip but failed to match on sourcePort" (where those things are the matcher names).

Definitely not a blocker though

File: docs/root/intro/arch_overview/advanced/matching/_include/listener_tls.yaml
Comment (kyessenov) [reply docs/root/intro/arch_overview/advanced/matching/_include/listener_tls.yaml:50]: There is a proposal https://github.com/cncf/xds/pull/28.

File: source/server/filter_chain_manager_impl.cc
Comment (kyessenov) [reply source/server/filter_chain_manager_impl.cc:575]: Does the listener tcp access log with filter chain name not show the chosen filter chain? Extra trace logging in the matcher framework would also be fine.

Comment (howardjohn) [reply source/server/filter_chain_manager_impl.cc:575]: That works for matching, I was thinking more about when it doesn't match. I suppose both are useful but would guess debugging non-match is more common.

I was playing with this PR earlier and got stuck on two issues (with my config, not the code here). Both ended up being pointing to filter chains that do not exist; so really the distinguishing of "no match" and "invalid action" would have been enough to help me debug the issues I have faced so far

Comment (kyessenov) [reply source/server/filter_chain_manager_impl.cc:575]: I see. Let me think about it. This was intended for an eventually consistent model where some filter chains might arrive later. Right now we process everything in a batch so can easily log stuff.

**Comment by kyessenov** [comment]
@phlax can you take a look again? I think this PR is waiting for your approval.

File: docs/root/intro/arch_overview/advanced/matching/matching_listener.rst
Comment (phlax) [reply docs/root/intro/arch_overview/advanced/matching/matching_listener.rst:32]: i think they are a helpful guide to the user - both visually that there is a full file that the snippet is taken from and for quickly finding the snippet in the config

i wont block on it tho as usage is inconsistent elsewhere

**Comment by phlax** [comment]
docs are rendered here https://storage.googleapis.com/envoy-pr/20110/docs/index.html

**Comment by phlax** [comment]
@howardjohn would you mind giving this a final pass/stamp

@htuch this one is awaiting an api approval

**Review by howardjohn — DISMISSED** [review]
I am not familiar enough with Envoy internals to review the code, but I imported this to our control plane and played around with it for a few hours, works really well

File: docs/root/intro/arch_overview/advanced/matching/_include/listener_complicated.yaml
Comment (htuch) [inline docs/root/intro/arch_overview/advanced/matching/_include/listener_complicated.yaml:78]: @kyessenov have you considered an input matcher that is basically just a CEL expression?

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by htuch** [comment]
@snowp might make sense for you to take a look from generic matcher perspective as well.

File: docs/root/intro/arch_overview/advanced/matching/_include/listener_complicated.yaml
Comment (kyessenov) [reply docs/root/intro/arch_overview/advanced/matching/_include/listener_complicated.yaml:78]: Yeah, https://github.com/cncf/xds/pull/31 is open.

Comment (htuch) [reply docs/root/intro/arch_overview/advanced/matching/_include/listener_complicated.yaml:78]: Awesome, let's continue there :)

**Review by phlax — APPROVED** [review]
lgtm from docs pov - thanks @kyessenov 

@snowp do you want to give this a final pass

/wait-any

File: source/server/filter_chain_manager_impl.cc
Comment (snowp) [inline source/server/filter_chain_manager_impl.cc:584]: Maybe add a trace/debug log here per @howardjohn's point on debugging?

File: api/envoy/config/listener/v3/listener.proto
Comment (snowp) [inline api/envoy/config/listener/v3/listener.proto:129]: What happens if both are defined?

File: docs/root/intro/arch_overview/advanced/matching/matching_listener.rst
Comment (snowp) [inline docs/root/intro/arch_overview/advanced/matching/matching_listener.rst:18]: Should this be exposed to the user in some way? ie fail the config or at the very least log something?

File: docs/root/intro/arch_overview/advanced/matching/matching_listener.rst
Comment (snowp) [inline docs/root/intro/arch_overview/advanced/matching/matching_listener.rst:54]: extraneous space after http

File: api/envoy/config/listener/v3/listener.proto
Comment (kyessenov) [reply api/envoy/config/listener/v3/listener.proto:129]: The filter chain match is ignored when listener matcher is defined. I added a debug log to warn on listener construction.

File: docs/root/intro/arch_overview/advanced/matching/matching_listener.rst
Comment (kyessenov) [reply docs/root/intro/arch_overview/advanced/matching/matching_listener.rst]: Thanks, fixed.

File: docs/root/intro/arch_overview/advanced/matching/matching_listener.rst
Comment (kyessenov) [reply docs/root/intro/arch_overview/advanced/matching/matching_listener.rst:18]: Added a debug log. Failing might complicate migration IMHO, so just ignoring the field seems reasonable as it is an opt-in feature.

File: source/server/filter_chain_manager_impl.cc
Comment (kyessenov) [reply source/server/filter_chain_manager_impl.cc:589]: Added a debug log when matcher refers to a missing chain. I think that's the confusing state that @howardjohn referred to.

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Comment by mattklein123** [comment]
@kyessenov I think we forget a release note for this. Do you mind doing a follow up PR to add a release note so that we can have that be part of the next release? Thank you.

**Pull request merged**
